src/builder.ts

```diff
@@ -62,64 +60,71 @@ export default class Builder {
 	 */
 	public createBuildStream(buildOpts: Object, hooks: Plugin.BuildHooks = {}, handler: ErrorHandler = emptyHandler): NodeJS.ReadWriteStream {
 
-		const self = this
+		const layers: string[] = [];
+		let errored = false;
 
-		this.layers = []
-
-		// Create a stream to be passed into the docker daemon
-		const inputStream = es.through()
+		// Create input/output streams to be passed to the docker daemon
+		const inputStream = new Stream.PassThrough();
+		const outputStream = new Stream.PassThrough();
 
 		// Create a bi-directional stream
 		const dup = duplexify()
-
-		// Connect the input stream to the rw stream
 		dup.setWritable(inputStream)
+		dup.setReadable(outputStream);
+
+		// Catch any errors the stream produces - dup.on('error') covers errors
+		// emitted by both inputStream and outputStream
+		dup.on('error', (err: Error) => {
+			this.callHook(hooks, 'buildFailure', handler, err, layers);
+		});
+
+		outputStream
+		.on('error', () => errored = true)
+		.on('end', () => {
+			if (!errored) {
+				this.callHook(hooks, 'buildSuccess', handler, _.last(layers), layers);
+			}
+		});
```

Comment:
This code block was moved up here as suggested by Alexis (more or less). It makes it a bit easier to read.